firedrake/slate/preconditioners.py

```diff
@@ -0,0 +1,285 @@
+        # TODO: Add support for mixed tensors and these variables
+        # become unnecessary
+        split_forms = split_form(new_form)
+        A = Tensor(next(sf.form for sf in split_forms
+                        if sf.indices == (0, 0)))
+        B = Tensor(next(sf.form for sf in split_forms
+                        if sf.indices == (1, 0)))
+        C = Tensor(next(sf.form for sf in split_forms
+                        if sf.indices == (1, 1)))
+        trial = TrialFunction(FunctionSpace(V.mesh(),
+                                            BrokenElement(W.ufl_element())))
+        K_local = Tensor(gammar('+') * ufl.dot(trial, n) * ufl.dS)
+
+        # Split functions and reconstruct each bit separately
+        sigma_h, u_h = self.broken_solution.split()
+        g, f = self.broken_rhs.split()
+
+        # Pressure reconstruction
+        M = B * A.inv * B.T + C
+        u_sol = M.inv * f + M.inv * (B * A.inv *
+                                     K_local.T * self.trace_solution
+                                     - B * A.inv * g)
+        self._assemble_pressure = create_assembly_callable(
+            u_sol,
+            tensor=u_h,
+            form_compiler_parameters=context.fc_params)
+
+        # Velocity reconstruction
+        sigma_sol = A.inv * g + A.inv * (B.T * u_h -
+                                         K_local.T * self.trace_solution)
+        self._assemble_velocity = create_assembly_callable(
+            sigma_sol,
+            tensor=sigma_h,
+            form_compiler_parameters=context.fc_params)
+
+        # Set up the projector for projecting the broken solution
+        # into the unbroken finite element spaces
+        # NOTE: Tolerance here matters!
+        sigma_b, _ = self.broken_solution.split()
+        sigma_u, _ = self.unbroken_solution.split()
+        self.projector = Projector(sigma_b,
+                                   sigma_u,
+                                   solver_parameters={"ksp_type": "cg",
+                                                      "ksp_rtol": 1e-13})
+
+    def update(self, pc):
+        """Update by assembling into the operator. No need to
+        reconstruct symbolic objects.
+        """
+        self._assemble_S()
+        self.S.force_evaluation()
+        self._assemble_Srhs()
+
+    def apply(self, pc, x, y):
+        """We solve the forward eliminated problem for the
+        approximate traces of the scalar solution (the multipliers)
+        and reconstruct the "broken flux and scalar variable."
+
+        Lastly, we project the broken solutions into the mimetic
+        non-broken finite element space.
+        """
+        from firedrake import project
+
+        # Transfer non-broken x into a firedrake function
+        with self.unbroken_rhs.dat.vec as v:
+            x.copy(v)
+
+        # Transfer unbroken_rhs into broken_rhs
+        field0, field1 = self.unbroken_rhs.split()
+        bfield0, bfield1 = self.broken_rhs.split()
```

Comment:
I think there are a few options:

1. default to `V*P` documenting (and via `view(pc)`)  with a solver option, say `-trace_field_order V,P` or `-trace_field_order P,V`

2. Require the user to specify (using same options)

3. Require use to specify (via `appctx`)

4. Determine automagically (sort of as you were doing, but only do it once!)